File: leetcode/Medium/python/189. Rotate Array.py
'''
https://leetcode.com/problems/rotate-array/description/
'''

class Solution:
    def rotate(self, nums, k):
        """
        Поворот массива вправо на k позиций
        
        Функция изменяет массив nums in-place, поворачивая его элементы вправо на k шагов.

        :param nums: список целых чисел
        :param k: число шагов для поворота массива
        :return: None
        """
        n = len(nums)
        k %= n

        def reverse(start, end):
            while start < end:
                nums[start], nums[end] = nums[end], nums[start]
                start += 1
                end -= 1

        reverse(0, n - 1)
        reverse(0, k - 1)
        reverse(k, n - 1)

''' Полезные ссылки: '''
# 1. 💠Telegram💠❃ Хижина программиста Æ: https://t.me/hut_programmer_07
# 2. 💠Telegram №1💠 @quadd4rv1n7
# 3. 💠Telegram №2💠 @dupley_maxim_1999
# 4. Rutube канал: https://rutube.ru/channel/4218729/
# 5. Plvideo канал: https://plvideo.ru/channel/AUPv_p1r5AQJ
# 6. YouTube канал: https://www.youtube.com/@it-coders
# 7. ВК группа: https://vk.com/science_geeks

File: leetcode/Medium/javascript/189. Rotate Array.js
/**
 * https://leetcode.com/problems/rotate-array/description/
 */

/**
 * Поворот массива вправо на k позиций
 * @param {number[]} nums - массив целых чисел
 * @param {number} k - число шагов для поворота массива
 * @return {void} Изменяет массив in-place
 */
var rotate = function(nums, k) {
    const n = nums.length;
    k %= n;

    const reverse = (start, end) => {
        while (start < end) {
            [nums[start], nums[end]] = [nums[end], nums[start]];
            start++;
            end--;
        }
    }

    reverse(0, n - 1);
    reverse(0, k - 1);
    reverse(k, n - 1);
};

/*
''' Полезные ссылки: '''
# 1. 💠Telegram💠❃ Хижина программиста Æ: https://t.me/hut_programmer_07
# 2. 💠Telegram №1💠 @quadd4rv1n7
# 3. 💠Telegram №2💠 @dupley_maxim_1999
# 4. Rutube канал: https://rutube.ru/channel/4218729/
# 5. Plvideo канал: https://plvideo.ru/channel/AUPv_p1r5AQJ
# 6. YouTube канал: https://www.youtube.com/@it-coders
# 7. ВК группа: https://vk.com/science_geeks
*/